From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 24 Jun 2016 17:09:35 -0400
Subject: dm error: add DAX support
References: bnc#1023798,FATE#321463
Patch-Mainline: v4.8-rc1
Git-commit: f8df1fdf18839cb4ef2035310bb9b6ec02025598

Allow the error target to replace an existing DAX-enabled target.

Signed-off-by: Mike Snitzer <snitzer@redhat.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/md/dm-table.c  |    3 ++-
 drivers/md/dm-target.c |    9 ++++++++-
 2 files changed, 10 insertions(+), 2 deletions(-)

--- a/drivers/md/dm-table.c
+++ b/drivers/md/dm-table.c
@@ -922,7 +922,8 @@ static int dm_table_determine_type(struc
 	if (bio_based) {
 		/* We must use this table as bio-based */
 		t->type = DM_TYPE_BIO_BASED;
-		if (dm_table_supports_dax(t))
+		if (dm_table_supports_dax(t) ||
+		    (list_empty(devices) && live_md_type == DM_TYPE_DAX_BIO_BASED))
 			t->type = DM_TYPE_DAX_BIO_BASED;
 		return 0;
 	}
--- a/drivers/md/dm-target.c
+++ b/drivers/md/dm-target.c
@@ -148,9 +148,15 @@ static void io_err_release_clone_rq(stru
 {
 }
 
+static long io_err_direct_access(struct dm_target *ti, sector_t sector,
+				 void **kaddr, pfn_t *pfn, long size)
+{
+	return -EIO;
+}
+
 static struct target_type error_target = {
 	.name = "error",
-	.version = {1, 4, 0},
+	.version = {1, 5, 0},
 	.features = DM_TARGET_WILDCARD,
 	.ctr  = io_err_ctr,
 	.dtr  = io_err_dtr,
@@ -158,6 +164,7 @@ static struct target_type error_target =
 	.map_rq = io_err_map_rq,
 	.clone_and_map_rq = io_err_clone_and_map_rq,
 	.release_clone_rq = io_err_release_clone_rq,
+	.direct_access = io_err_direct_access,
 };
 
 int __init dm_target_init(void)
